// nestjs-zod/dist/dto-XwCwaTfM.cjs
//#region rolldown:runtime
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") {
		for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
			key = keys[i];
			if (!__hasOwnProp.call(to, key) && key !== except) {
				__defProp(to, key, {
					get: ((k) => from[k]).bind(null, key),
					enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
				});
			}
		}
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
let zod_v4_core = require("zod/v4/core");
let deepmerge = require("deepmerge");
deepmerge = __toESM(deepmerge);
let zod_v3 = require("zod/v3");

//#region src/assert.ts
function assert(condition, message = "Assertion failed") {
	if (!condition) throw new Error(`[nestjs-zod] ${message}`);
}

//#endregion
//#region src/const.ts
const PREFIX = "x-nestjs_zod";
const EMPTY_TYPE_KEY = `${PREFIX}-empty-type`;
const DEFS_KEY = `${PREFIX}-$defs`;
const PARENT_ID_KEY = `${PREFIX}-parent-id`;
const PARENT_TITLE_KEY = `${PREFIX}-parent-title`;
const PARENT_ADDITIONAL_PROPERTIES_KEY = `${PREFIX}-parent-additional-properties`;
const PARENT_HAS_REFS_KEY = `${PREFIX}-parent-has-refs`;
const UNWRAP_ROOT_KEY = `${PREFIX}-unwrap-root`;
const HAS_NULL_KEY = `${PREFIX}-has-null`;
const HAS_CONST_KEY = `${PREFIX}-has-const`;
const SELF_REQUIRED_KEY = `${PREFIX}-self-required`;

//#endregion
//#region src/utils.ts
function fixAllRefs({ schema, defRenames, rootSchemaName }) {
	return walkJsonSchema(schema, (s) => {
		if (s.$ref) {
			if (s.$ref.startsWith("#/$defs/")) {
				const oldDefName = s.$ref.replace("#/$defs/", "");
				const newDefName = defRenames?.[oldDefName];
				if (newDefName) s.$ref = `#/$defs/${newDefName}`;
			}
			s.$ref = s.$ref.replace("#/$defs/", "#/components/schemas/");
			if (s.$ref === "#") {
				if (!rootSchemaName) throw new Error("[fixAllRefs] rootSchemaName is required when fixing a ref to #");
				s.$ref = `#/components/schemas/${rootSchemaName}`;
			}
		}
		return s;
	}, { clone: true });
}
/**
* By default, zod generates openapi schemas that are compatible with OpenAPI
* 3.1.  But OpenAPI 3.0 supports a weird flavour of JSONSchema they call the
* "subset superset"
*
* This function converts the schema to the OpenAPI 3.0 subset superset format.
*
* See more information here:
* https://www.apimatic.io/blog/2021/09/migrating-to-and-from-openapi-3-1
*/
function convertToOpenApi3Point0(schema) {
	return walkJsonSchema(schema, (s) => {
		if (Object.keys(s).length === 1 && s.anyOf) {
			const nullSchema = s.anyOf.findIndex((subSchema) => subSchema.type === "null");
			if (nullSchema === -1) return s;
			s.anyOf.splice(nullSchema, 1);
			if (s.anyOf.length === 1) return {
				...s.anyOf[0],
				nullable: true
			};
			return { anyOf: s.anyOf.map((subSchema) => ({
				...subSchema,
				nullable: true
			})) };
		}
		if (s.const) {
			s.enum = [s.const];
			delete s.const;
		}
		return s;
	}, { clone: true });
}
function walkJsonSchema(schema, callback, options) {
	schema = callback(options?.clone ? (0, deepmerge.default)(schema, {}) : schema);
	if (schema.type === "object" && schema.properties) for (const key in schema.properties) schema.properties[key] = walkJsonSchema(schema.properties[key], callback);
	if (schema.type === "array" && Array.isArray(schema.items)) schema.items = schema.items.map((item) => walkJsonSchema(item, callback));
	if (schema.type === "array" && schema.items) schema.items = walkJsonSchema(schema.items, callback);
	if (schema.oneOf) schema.oneOf = schema.oneOf.map((subSchema) => walkJsonSchema(subSchema, callback));
	if (schema.anyOf) schema.anyOf = schema.anyOf.map((subSchema) => walkJsonSchema(subSchema, callback));
	if (schema.allOf) schema.allOf = schema.allOf.map((subSchema) => walkJsonSchema(subSchema, callback));
	if (typeof schema.additionalProperties === "object") schema.additionalProperties = walkJsonSchema(schema.additionalProperties, callback);
	return schema;
}

//#endregion
//#region src/zodV3ToOpenApi.ts
function is(input, factory) {
	return factory === zod_v3.z[input._def.typeName];
}
/**
* @deprecated `zodToOpenAPI` will be removed in a future version, since zod
* v4 adds built-in support for generating OpenAPI schemas from zod schemas.
*/
function zodV3ToOpenAPI(zodType, visited = /* @__PURE__ */ new Set()) {
	const object = {};
	if (zodType.description) object.description = zodType.description;
	if (is(zodType, zod_v3.z.ZodString)) {
		const { checks } = zodType._def;
		object.type = "string";
		for (const check of checks) if (check.kind === "min") object.minLength = check.value;
		else if (check.kind === "max") object.maxLength = check.value;
		else if (check.kind === "email") object.format = "email";
		else if (check.kind === "url") object.format = "uri";
		else if (check.kind === "uuid") object.format = "uuid";
		else if (check.kind === "cuid") object.format = "cuid";
		else if (check.kind === "regex") object.pattern = check.regex.source;
		else if (check.kind === "datetime") object.format = "date-time";
	}
	if (is(zodType, zod_v3.z.ZodBoolean)) object.type = "boolean";
	if (is(zodType, zod_v3.z.ZodNumber)) {
		const { checks } = zodType._def;
		object.type = "number";
		for (const check of checks) if (check.kind === "int") object.type = "integer";
		else if (check.kind === "min") {
			object.minimum = check.value;
			object.exclusiveMinimum = !check.inclusive;
		} else if (check.kind === "max") {
			object.maximum = check.value;
			object.exclusiveMaximum = !check.inclusive;
		} else if (check.kind === "multipleOf") object.multipleOf = check.value;
	}
	if (is(zodType, zod_v3.z.ZodBigInt)) {
		object.type = "integer";
		object.format = "int64";
	}
	if (is(zodType, zod_v3.z.ZodArray)) {
		const { minLength, maxLength, type } = zodType._def;
		object.type = "array";
		if (minLength) object.minItems = minLength.value;
		if (maxLength) object.maxItems = maxLength.value;
		object.items = zodV3ToOpenAPI(type, visited);
	}
	if (is(zodType, zod_v3.z.ZodTuple)) {
		const { items } = zodType._def;
		object.type = "array";
		object.items = { oneOf: items.map((item) => zodV3ToOpenAPI(item, visited)) };
	}
	if (is(zodType, zod_v3.z.ZodSet)) {
		const { valueType, minSize, maxSize } = zodType._def;
		object.type = "array";
		if (minSize) object.minItems = minSize.value;
		if (maxSize) object.maxItems = maxSize.value;
		object.items = zodV3ToOpenAPI(valueType, visited);
		object.uniqueItems = true;
	}
	if (is(zodType, zod_v3.z.ZodUnion)) {
		const { options } = zodType._def;
		object.oneOf = options.map((option) => zodV3ToOpenAPI(option, visited));
	}
	if (is(zodType, zod_v3.z.ZodDiscriminatedUnion)) {
		const { options } = zodType._def;
		object.oneOf = [];
		for (const schema of options.values()) object.oneOf.push(zodV3ToOpenAPI(schema, visited));
	}
	if (is(zodType, zod_v3.z.ZodLiteral)) {
		const { value } = zodType._def;
		if (typeof value === "string") {
			object.type = "string";
			object.enum = [value];
		}
		if (typeof value === "number") {
			object.type = "number";
			object.minimum = value;
			object.maximum = value;
		}
		if (typeof value === "boolean") object.type = "boolean";
	}
	if (is(zodType, zod_v3.z.ZodEnum)) {
		const { values } = zodType._def;
		object.type = "string";
		object.enum = values;
	}
	if (is(zodType, zod_v3.z.ZodNativeEnum)) {
		const { values } = zodType._def;
		object.type = "string";
		object.enum = Object.values(values);
		object["x-enumNames"] = Object.keys(values);
	}
	if (is(zodType, zod_v3.z.ZodTransformer)) {
		const { schema } = zodType._def;
		Object.assign(object, zodV3ToOpenAPI(schema, visited));
	}
	if (is(zodType, zod_v3.z.ZodNullable)) {
		const { innerType } = zodType._def;
		Object.assign(object, zodV3ToOpenAPI(innerType, visited));
		object.nullable = true;
	}
	if (is(zodType, zod_v3.z.ZodOptional)) {
		const { innerType } = zodType._def;
		Object.assign(object, zodV3ToOpenAPI(innerType, visited));
	}
	if (is(zodType, zod_v3.z.ZodDefault)) {
		const { defaultValue, innerType } = zodType._def;
		Object.assign(object, zodV3ToOpenAPI(innerType, visited));
		object.default = defaultValue();
	}
	if (is(zodType, zod_v3.z.ZodObject)) {
		const { shape } = zodType._def;
		object.type = "object";
		object.properties = {};
		object.required = [];
		for (const [key, schema] of Object.entries(shape())) {
			object.properties[key] = zodV3ToOpenAPI(schema, visited);
			if (![zod_v3.z.ZodOptional.name, zod_v3.z.ZodDefault.name].includes(schema.constructor.name)) object.required.push(key);
		}
		if (object.required.length === 0) delete object.required;
	}
	if (is(zodType, zod_v3.z.ZodRecord)) {
		const { valueType } = zodType._def;
		object.type = "object";
		object.additionalProperties = zodV3ToOpenAPI(valueType, visited);
	}
	if (is(zodType, zod_v3.z.ZodIntersection)) {
		const { left, right } = zodType._def;
		const merged = (0, deepmerge.default)(zodV3ToOpenAPI(left, visited), zodV3ToOpenAPI(right, visited), { arrayMerge: (target, source) => {
			const mergedSet = new Set([...target, ...source]);
			return Array.from(mergedSet);
		} });
		Object.assign(object, merged);
	}
	if (is(zodType, zod_v3.z.ZodEffects)) {
		const { schema } = zodType._def;
		Object.assign(object, zodV3ToOpenAPI(schema, visited));
	}
	if (is(zodType, zod_v3.z.ZodLazy)) {
		const { getter } = zodType._def;
		if (visited.has(getter)) return object;
		visited.add(getter);
		Object.assign(object, zodV3ToOpenAPI(getter(), visited));
	}
	return object;
}

//#endregion
//#region src/symbols.ts
const ioSymbol = Symbol("io");

//#endregion
//#region src/dto.ts
function createZodDto(schema, options) {
	class AugmentedZodDto {
		static {
			this.isZodDto = true;
		}
		static {
			this.schema = schema;
		}
		static {
			this.codec = options?.codec || false;
		}
		static {
			this[ioSymbol] = "input";
		}
		static create(input) {
			return this.schema.parse(input);
		}
		static get Output() {
			assert("_zod" in schema, "Output DTOs can only be created from zod v4 schemas");
			class AugmentedZodDto$1 {
				static {
					this.isZodDto = true;
				}
				static {
					this.schema = schema;
				}
				static {
					this[ioSymbol] = "output";
				}
				static create(input) {
					return this.schema.parse(input);
				}
				static _OPENAPI_METADATA_FACTORY() {
					return openApiMetadataFactory({
						schema: this.schema,
						io: "output"
					});
				}
			}
			Object.defineProperty(AugmentedZodDto$1, "name", { value: `${this.name}_Output` });
			return AugmentedZodDto$1;
		}
		static _OPENAPI_METADATA_FACTORY() {
			return openApiMetadataFactory({
				schema: this.schema,
				io: "input"
			});
		}
	}
	return AugmentedZodDto;
}
function openApiMetadataFactory({ schema, io }) {
	if (!("_zod" in schema) && "_def" in schema && io === "output") throw new Error("[nestjs-zod] Output schemas are not supported for zod@v3");
	if (!("_zod" in schema) && !("_def" in schema)) return {};
	const { $defs, $schema, ...generatedJsonSchema } = generateJsonSchema(schema, io);
	/**
	* nestjs expects us to return a record of properties
	*
	* However, in some cases, we can't return a record of properties.  For
	* example, arrays, intersections, and unions can not be represented like this
	*
	* As a workaround, we wrap the schema in a "root" object.  Then in the
	* `cleanupOpenApiDoc` function, we unwrap the root object.
	*/
	const jsonSchema = !isObjectTypeWithProperties(generatedJsonSchema) ? {
		type: "object",
		properties: { root: {
			...generatedJsonSchema,
			[UNWRAP_ROOT_KEY]: true
		} },
		$defs
	} : {
		...generatedJsonSchema,
		$defs
	};
	const { hasRefs, hasNull, hasConst } = getSchemaMetadata(jsonSchema);
	let properties = {};
	for (let [propertyKey, propertySchema] of Object.entries(jsonSchema.properties || {})) {
		const newPropertySchema = {
			...propertySchema,
			type: propertySchema.type || ""
		};
		if (hasConst) newPropertySchema[HAS_CONST_KEY] = true;
		if (hasNull) newPropertySchema[HAS_NULL_KEY] = true;
		if (hasRefs) newPropertySchema[PARENT_HAS_REFS_KEY] = true;
		if (typeof propertySchema.type !== "string") newPropertySchema[EMPTY_TYPE_KEY] = true;
		const required = Boolean("required" in jsonSchema && jsonSchema.required?.includes(propertyKey));
		if (newPropertySchema.type === "object") {
			newPropertySchema.selfRequired = required;
			newPropertySchema[SELF_REQUIRED_KEY] = required;
		} else newPropertySchema.required = required;
		if (jsonSchema.$defs) newPropertySchema[DEFS_KEY] = jsonSchema.$defs;
		if (jsonSchema.id) newPropertySchema[PARENT_ID_KEY] = jsonSchema.id;
		if (jsonSchema.title) newPropertySchema[PARENT_TITLE_KEY] = jsonSchema.title;
		if (typeof jsonSchema.additionalProperties === "boolean") newPropertySchema[PARENT_ADDITIONAL_PROPERTIES_KEY] = jsonSchema.additionalProperties;
		properties[propertyKey] = newPropertySchema;
	}
	return properties;
}
function generateJsonSchema(schema, io) {
	const generatedJsonSchema = "_zod" in schema ? (0, zod_v4_core.toJSONSchema)(schema, {
		io,
		override: ({ jsonSchema }) => {
			if (io === "output" && "id" in jsonSchema) jsonSchema.id = `${jsonSchema.id}_Output`;
		}
	}) : zodV3ToOpenAPI(schema);
	const $defs = "$defs" in generatedJsonSchema && generatedJsonSchema.$defs ? generatedJsonSchema.$defs : void 0;
	const newSchema = fixRefsToPointById(generatedJsonSchema, $defs);
	const newDefs = {};
	Object.entries($defs || {}).forEach(([defKey, defValue]) => {
		if (defValue.id) {
			const newKey = defValue.id || defKey;
			if (newDefs[newKey]) throw new Error(`[nestjs-zod] Duplicate id in $defs: ${newKey}`);
			newDefs[newKey] = fixRefsToPointById(defValue, $defs);
		} else newDefs[defKey] = fixRefsToPointById(defValue, $defs);
	});
	if ($defs) newSchema.$defs = newDefs;
	return newSchema;
}
/**
* Changes all $refs in a schema to point based off the schema's ID, not the
* keys of the $defs object.  For example, given this schema:
* ```json
* {
*  type: "object",
*  properties: {
*    author: {
*      $ref: "#/$defs/Author"
*    }
*  }
* }
* ```
* and this $defs object:
* ```json
* {
*   Author: {
*    id: "Author_Output",
*    type: "object",
*    properties: {
*      // ...
*    }
*  }
* }
* ```
* It will return this schema:
* ```json
* {
*  type: "object",
*  properties: {
*    author: {
*      $ref: "#/$defs/Author_Output"
*    }
*  }
* }
* ```
*/
function fixRefsToPointById(rootSchema, $defs) {
	return walkJsonSchema(rootSchema, (schema) => {
		if (schema.$ref && schema.$ref.startsWith("#/$defs/")) {
			const defKey = schema.$ref.replace("#/$defs/", "");
			const defId = $defs?.[defKey].id;
			if (defId) schema.$ref = `#/$defs/${defId}`;
		}
		return schema;
	}, { clone: true });
}
function getSchemaMetadata(jsonSchema) {
	let hasRefs = false;
	let hasNull = false;
	let hasConst = false;
	walkJsonSchema(jsonSchema, (schema) => {
		if (schema.type === "null") hasNull = true;
		if (schema.$ref) hasRefs = true;
		if (schema.const) hasConst = true;
		return schema;
	});
	return {
		hasRefs,
		hasNull,
		hasConst
	};
}
function isZodDto(metatype) {
	return Boolean(metatype && (typeof metatype === "object" || typeof metatype === "function") && "isZodDto" in metatype && metatype.isZodDto);
}
function isObjectTypeWithProperties(jsonSchema) {
	return jsonSchema.type === "object" && !!jsonSchema.properties && Object.keys(jsonSchema.properties).length > 0;
}

//#endregion
Object.defineProperty(exports, 'DEFS_KEY', {
  enumerable: true,
  get: function () {
    return DEFS_KEY;
  }
});
Object.defineProperty(exports, 'EMPTY_TYPE_KEY', {
  enumerable: true,
  get: function () {
    return EMPTY_TYPE_KEY;
  }
});
Object.defineProperty(exports, 'HAS_CONST_KEY', {
  enumerable: true,
  get: function () {
    return HAS_CONST_KEY;
  }
});
Object.defineProperty(exports, 'HAS_NULL_KEY', {
  enumerable: true,
  get: function () {
    return HAS_NULL_KEY;
  }
});
Object.defineProperty(exports, 'PARENT_ADDITIONAL_PROPERTIES_KEY', {
  enumerable: true,
  get: function () {
    return PARENT_ADDITIONAL_PROPERTIES_KEY;
  }
});
Object.defineProperty(exports, 'PARENT_HAS_REFS_KEY', {
  enumerable: true,
  get: function () {
    return PARENT_HAS_REFS_KEY;
  }
});
Object.defineProperty(exports, 'PARENT_ID_KEY', {
  enumerable: true,
  get: function () {
    return PARENT_ID_KEY;
  }
});
Object.defineProperty(exports, 'PARENT_TITLE_KEY', {
  enumerable: true,
  get: function () {
    return PARENT_TITLE_KEY;
  }
});
Object.defineProperty(exports, 'SELF_REQUIRED_KEY', {
  enumerable: true,
  get: function () {
    return SELF_REQUIRED_KEY;
  }
});
Object.defineProperty(exports, 'UNWRAP_ROOT_KEY', {
  enumerable: true,
  get: function () {
    return UNWRAP_ROOT_KEY;
  }
});
Object.defineProperty(exports, '__toESM', {
  enumerable: true,
  get: function () {
    return __toESM;
  }
});
Object.defineProperty(exports, 'assert', {
  enumerable: true,
  get: function () {
    return assert;
  }
});
Object.defineProperty(exports, 'convertToOpenApi3Point0', {
  enumerable: true,
  get: function () {
    return convertToOpenApi3Point0;
  }
});
Object.defineProperty(exports, 'createZodDto', {
  enumerable: true,
  get: function () {
    return createZodDto;
  }
});
Object.defineProperty(exports, 'fixAllRefs', {
  enumerable: true,
  get: function () {
    return fixAllRefs;
  }
});
Object.defineProperty(exports, 'ioSymbol', {
  enumerable: true,
  get: function () {
    return ioSymbol;
  }
});
Object.defineProperty(exports, 'isZodDto', {
  enumerable: true,
  get: function () {
    return isZodDto;
  }
});
Object.defineProperty(exports, 'zodV3ToOpenAPI', {
  enumerable: true,
  get: function () {
    return zodV3ToOpenAPI;
  }
});
